### PR TITLE
Upgrade tiny mce

### PR DESCRIPTION
**Pull request by pparke**
Upgrade to TinyMCE 5. Allow external JavaScript URLs to be added. Add some custom JavaScript to remove extra line breaks and styles while preserving highlighted text.

### CODE REVIEW

**Comment by pparke** [comment]
https://app.shortcut.com/silverorangeinc/story/8181/update-admin-to-tinymce

File: Swat/SwatExternalJavaScriptHtmlHeadEntry.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+/**
+ * Stores and outputs an HTML head entry for a JavaScript include
+ *
+ * @package   Swat
+ * @copyright 2006-2016 silverorange
```
Comment (nburka) [inline Swat/SwatExternalJavaScriptHtmlHeadEntry.php:7]:
```suggestion
 * @copyright 2022 silverorange
```

File: Swat/SwatExternalJavaScriptHtmlHeadEntry.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+/**
+ * Stores and outputs an HTML head entry for a JavaScript include
```
Comment (nburka) [inline Swat/SwatExternalJavaScriptHtmlHeadEntry.php:4]:
```suggestion
 * Stores and outputs an HTML head entry for an external JavaScript resource
```

File: Swat/SwatExternalJavaScriptHtmlHeadEntry.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        );
+    }
+
+    // }}}
+    // {{{ protected function displayInlineInternal()
+
+    protected function displayInlineInternal($path)
+    {
+        echo '<script type="text/javascript">';
+        readfile($path . $this->getUri());
```
Comment (nburka) [inline Swat/SwatExternalJavaScriptHtmlHeadEntry.php:30]: does this even work with an external resource? Even if it does, it'll be slow because it has to download the file from the server and then out it here.

I think this is just used by the HTML concentrate stuff that combines JS files into one big file. For external resources we shouldn't be doing that anyway. I think it would make sense to just return an empty string here instead and add a comment explaining why.

File: Swat/SwatTextareaEditor.php [hunk 4]
```diff
@@ -21,6 +21,8 @@ class SwatTextareaEditor extends SwatTextarea
     // }}}
     // {{{ public properties
 
+    public static $tiny_mce_api_key = '';
```
Comment (nburka) [inline Swat/SwatTextareaEditor.php:24]: I think Mike would prefer a protected static property and then a public static method to set that. Can you do that instead and update the site code to call `SwatTextareaEditor::setTinyMceApiKey('x');`?

File: Swat/SwatTextareaEditor.php [hunk 5]
```diff
@@ -335,7 +340,7 @@ protected function getInlineJavaScript()
         $lines = array();
         foreach ($this->getConfig() as $name => $value) {
             if (is_string($value)) {
-                $value = SwatString::quoteJavaScriptString($value);
+                $value = "'{$value}'";
```
Comment (nburka) [inline Swat/SwatTextareaEditor.php:343]: why did you remove `SwatString::quoteJavaScriptString`? That will escape any characters and avoid the possibility of JS injection.

File: Swat/SwatTextareaEditor.php [hunk 6]
```diff
@@ -349,6 +354,51 @@ protected function getInlineJavaScript()
         // Make removeformat button also clear inline alignments, styles,
```
Comment (nburka) [inline Swat/SwatTextareaEditor.php:354]: this comment looks out of date. Can you update it and also include a description of what you're doing in the post-paste step?

File: dependencies/swat.yaml [hunk 7]
```diff
@@ -75,6 +75,7 @@ Swat:
     packages/swat/javascript/swat-view.js:
     packages/swat/javascript/swat-z-index-manager.js:
     packages/swat/javascript/tiny_mce/tiny_mce.js:
+    packages/swat/javascript/tinymce/tinymce.min.js:
```
Comment (nburka) [inline dependencies/swat.yaml:78]: is this correct? I thought you were using an externally hosted JS for tinymce?

**Review by nburka — CHANGES_REQUESTED** [review]
There's also the discussion in Slack about allowing no API key, or a free API key, and then not showing using the paid-plugins in that case. https://silverorange.slack.com/archives/C041VKQTV/p1650396947501649?thread_ts=1650395367.499639&cid=C041VKQTV

File: Swat/SwatTextareaEditor.php (hunk 4 above)
Comment (pparke) [reply Swat/SwatTextareaEditor.php:24]: His comment said a method or property was ok https://silverorange.slack.com/archives/CBSTRPSPP/p1649982320793889?thread_ts=1649962756.957899&cid=CBSTRPSPP

File: Swat/SwatTextareaEditor.php (hunk 5 above)
Comment (pparke) [reply Swat/SwatTextareaEditor.php]: I think I removed it when I was trying something out that required some inline javascript and this was turning it into a string, I've changed it back

File: Swat/SwatTextareaEditor.php [hunk 8]
```diff
@@ -141,11 +141,17 @@ public function __construct($id = null)
         $this->requires_id = true;
         $this->rows = 30;
 
-        if (self::$tiny_mce_api_key !== null && !empty(self::$tiny_mce_api_key)) {
-            $tiny_mce_url = "https://cdn.tiny.cloud/1/".
-                self::$tiny_mce_api_key."/tinymce/5/tinymce.min.js";
+        if (
+            self::$tiny_mce_api_key !== null &&
```
Comment (nburka) [inline Swat/SwatTextareaEditor.php:145]: `empty(null)` returns true, so technically you don't need this. Typically though, we avoid using `empty()` and just use `=== null`. You're doing this in case there's an empty string, right? Can't we avoid that ever by just using the default config value of null?

File: Swat/SwatTextareaEditor.php [hunk 9]
```diff
@@ -262,11 +268,12 @@ protected function getConfig()
...
         $image_server = $this->image_server ? $this->image_server : '';
 
-        $has_api_key = (self::$tiny_mce_api_key !== null && !empty(self::$tiny_mce_api_key));
+        $has_api_key =
+            self::$tiny_mce_api_key !== null && !empty(self::$tiny_mce_api_key);
         $paste_plugin = $has_api_key ? ' powerpaste' : ' paste';
 
         $config = array(
-            'selector' => '#'.$this->id,
+            'selector' => '#' . $this->id,
```
Comment (nburka) [inline Swat/SwatTextareaEditor.php:276]: did prettier add this space? Typically we just do `'.$var`

**Review by nburka — CHANGES_REQUESTED** [review]

File: Swat/SwatTextareaEditor.php (hunk 9 above)
Comment (pparke) [reply Swat/SwatTextareaEditor.php:276]: prettier did
